src/colmap/util/threading_ctrl.h

```diff
@@ -0,0 +1,259 @@
+ protected:
+  // Register a new callback. Note that only registered callbacks can be
+  // set/reset and called from within the thread. Hence, this method should be
+  // called from the derived thread constructor.
+  void RegisterCallback(int id);
+
+  // Call back to the function with the specified name, if it exists.
+  void Callback(int id) const;
+
+  // Signal that the thread is setup. Only call this function once.
+  void SignalValidSetup();
+  void SignalInvalidSetup();
+
+ private:
+  bool setup_ = false;
+  bool setup_valid_ = false;
+  // list of callbacks
+  std::unordered_map<int, std::list<std::function<void()>>> callbacks_;
+  // check_if_stop function
+  std::function<bool()> check_if_stopped_fn;
+};
+
+// Helper class to create single threads with simple controls
+// Similar usage as ``Thread`` class in util/threading.h except for
+// initialization. e.g.,
+//
+// std::shared_ptr<Controller> controller = std::make_shared<Controller>(args);
+// std::unique_ptr<ControllerThread<Controller>> thread =
+// std::make_unique<ControllerThread<Controller>>(controller);
+//
+//
+template <class Controller>
+class ControllerThread {
+  // check if the Controller class is inherited from BaseController
+  static_assert(std::is_base_of<BaseController, Controller>::value,
+                "The controller needs to be inherited from BaseController");
+
+ public:
+  struct ThreadStatus {
+    // Check the status
+    bool IsStarted() const { return started; }
+    bool IsStopped() const { return stopped; }
+    bool IsPaused() const { return paused; }
+    bool IsRunning() const { return started && !pausing && !finished; }
+    bool IsFinished() const { return finished; }
+
+    // Update the status
+    void Start() {
+      started = true;
+      stopped = false;
+      paused = false;
+      pausing = false;
+      finished = false;
+    }
+
+    bool started = false;
+    bool stopped = false;
+    bool paused = false;
+    bool pausing = false;
+    bool finished = false;
+  };
+
+  // Check the state of the thread.
+  bool IsStarted() const { return status_.IsStarted(); }
+  bool IsStopped() const { return status_.IsStopped(); }
+  bool IsPaused() const { return status_.IsPaused(); }
+  bool IsRunning() const { return status_.IsRunning(); }
+  bool IsFinished() const { return status_.IsFinished(); }
+
+  explicit ControllerThread(std::shared_ptr<Controller> controller) {
+    controller_ = controller;
+    controller_->AddCallback(BaseController::FINISHED_CALLBACK,
+                             [&]() { status_.finished = true; });
+    controller_->AddCallback(BaseController::LOCK_MUTEX_CALLBACK, [&]() {
+      std::unique_lock<std::mutex> lock(mutex_);
+    });
+    controller_->AddCallback(BaseController::BLOCK_IF_PAUSED_CALLBACK, [&]() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      if (status_.paused) {
+        status_.pausing = true;
+        pause_condition_.wait(lock);
+        status_.pausing = false;
+      }
+    });
```

Comment:
This is redundant with `CheckIfStoppedFunc`, no?

Reply:
Same to the previous one. If we combine them together this is not needed.

Reply:
Yes this is not needed after merging.